### PR TITLE
A Semaphore for Parallelizable Reentrancy Protection

### PR DESCRIPTION
**Pull request by SergioDemianLerner**

### CODE REVIEW

File: EIPS/eip-5283.md
Comment (MicahZoltu) [inline EIPS/eip-5283.md:72]: Instead of linking to these, you should briefly describe their behavior.  This EIP won't be able to move to final until all of the EIPs it references are also final, and since this EIP is "competing" with these you certainly wouldn't want to be blocked by them not being included.

**Review by MicahZoltu — CHANGES_REQUESTED** [review]

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:19]:
```suggestion
The typical smart contract RPG uses a contract storage cell. The algorithm is simple: the code checks that a storage cell is 0 (or any other predefined constant) on entry, aborting if not, and then sets it to 1. After executing the required code, it resets the cell back to 0 before exiting. This is the algorithm implemented in OpenZeppelin's ReentrancyGuard. The algorithm results in a read-write pattern on the RPG's storage cell. This pattern prevents the parallelization of the execution of the smart contract for all known designs that try to provide fine-grained parallelization (detecting conflicts at the storage cell level). 
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:31]:
```suggestion
Several EVM-based blockchains have successfully tested designs for the parallelization of the EVM. The best results have been obtained with fine-grained parallelization where conflicts are detected by tracking writes and reads of individual storage cells. The designs based on tracking the use of accounts or contracts provide only minor benefits because most transactions use the same [EIP-20](./eip-20.md) contracts.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:80]:
```suggestion
The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost (such as [EIP-1109](./eip-1109.md) or [EIP-2046](./eip-2046.md)) is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with EIP-2046.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:100]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:10]:
```suggestion
requires: 20, 1109, 1153, 1283, 1352, 2046
---
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: EIPS/eip-5283.md
Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:72]: Ok, I'll describe them. But I can still mention the IP numbers right ?

Comment (MicahZoltu) [reply EIPS/eip-5283.md:72]: Not unless you want to be blocked by them.  Until an EIP is final, it is completely mutable.  If you refer to EIP-X today and then your EIP becomes final but EIP-X is not, then it is possible that EIP-X will be radically changed before it is finally merged as final and now your EIP doesn't make sense because it is referring to a document that is not the same as when you wrote the reference.

File: EIPS/eip-5283.md
Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:10]: Should an EIP referenced be in the "requires" list even if it doesn't need to be final for this EIP to be finalized ? For example, should competing proposals be referenced in "requires" ?

Comment (MicahZoltu) [reply EIPS/eip-5283.md:10]: You generally shouldn't include competing proposals in your EIP.  An EIP should stand on its own as much as possible, and external references should be to final EIPs (or EIPs that are expected to be final before your EIP makes it to final).  Keep in mind, the EIP process is for standardization and isn't an appropriate place to argue for/against this standard vs that standard.  Such material is better suited for a blog, article, docs site, discussions-to link, forum thread, etc.

My usual recommendation to authors is to remove all references to other EIPs that aren't required reading for understanding this standard, and if you want to mention something in the rationale as "we thought about doing X but decided against it because Y" then just briefly describe X rather than linking to a non-final EIP.

Keep in mind, EIPs can be completely/radically changed prior to becoming final so if you write your EIP and reference a non-final EIP then it is possible that the content you referenced may no longer make sense, which is why we don't allow non-final references in final EIPs.

File: EIPS/eip-5283.md
Comment (SamWilsn) [inline EIPS/eip-5283.md:33]: rage?

File: EIPS/eip-5283.md
Comment (SamWilsn) [inline EIPS/eip-5283.md:31]: These paragraphs read more like Motivation than Rationale. The Motivation section describes why the EIP as a whole needs to exist, while the Rationale section should describe individual technical choices made within the EIP.

File: EIPS/eip-5283.md
Comment (Pandapip1) [reply EIPS/eip-5283.md:33]: 💢

Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:33]: :)

File: EIPS/eip-5283.md
Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:72]: Understood!

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:8]:
```suggestion
type: Standards Track
category: ERC
```
```suggestion
type: Informational
```
This EIP is not a core EIP. Please pick one of these (I'm inclined towards the latter).

Comment (MicahZoltu) [reply EIPS/eip-5283.md:8]: Unless I'm reading this EIP wrong, it *is* a Core EIP as it introduces a precompile to the EVM.

Comment (Pandapip1) [reply EIPS/eip-5283.md:8]: Disregard my previous comment.

File: EIPS/eip-5283.md
Comment (moodysalem) [inline EIPS/eip-5283.md:74]: why doesn't this help with parallelization? transactions that use any amount of transient storage can be executed in parallel, even if they use the same slot

Comment (moodysalem) [reply EIPS/eip-5283.md:74]: is this about parallel execution for a single transaction?

Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:74]: My mistake! I misplaced that sentence when rewritting the paragraph according to the last review.

Comment (moodysalem) [reply EIPS/eip-5283.md]: so i understand correctly, the transient storage eip provides the same benefits as this eip, and is more general, i.e. not limited to reentrancy locks? would be helpful to have a `discussions-to` url in the header early so there's a better place to discuss, but also think it should be explained in the alternate or rationale how this eip is different

**Comment by MicahZoltu** [comment]
This cannot be merged as a draft until there is a `discussions-to` with a link to a thread on Ethereum Magicians.

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:1]: Missing discussions-to.

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:104]:
```suggestion

A call to second() directly from a transaction does not revert, but a call to first() does revert.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:106]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:91]:
```suggestion

This change requires a hard fork and therefore all full nodes must be updated.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:85]:
```suggestion

A gas cost of 100 represents a worst-case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.

Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in total).
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:88]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:12]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:25]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:30]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:33]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:64]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:80]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:83]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:109]:
```suggestion
Needs discussion.
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

File: EIPS/eip-5283.md
Comment (SergioDemianLerner) [reply EIPS/eip-5283.md]: yes, transient opcodes are better from the EVM "purity" standpoint. But I haven't heard of any use for the transient opcodes apart from mutexes. I will add the discussions tag, where we should be chatting.

File: EIPS/eip-5283.md
Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:99]: Yul is ok?

File: EIPS/eip-5283.md
Comment (SergioDemianLerner) [reply EIPS/eip-5283.md:106]: No security risk has been identified in recent discussions.

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:106]:
```suggestion
No security risks have been identified related to this change.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:36]:
```suggestion
```solidity
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:71]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:73]:
```suggestion
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:102]:
```suggestion
A call to `second()` directly from a transaction does not revert, but a call to `first()` does revert.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:106]:
```suggestion
No security risks have been identified related to this change.
```

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:109]:
```suggestion

Copyright and related rights waived via [CC0](../LICENSE.md).
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
Just some minor changes, and then I think this is ready.

File: EIPS/eip-5283.md
Comment (xinbenlv) [inline EIPS/eip-5283.md:28]: A precompiled contract usually takes time (month to years) to converge to a consensus. I suggest use a placeholder for `0x0A` and finalize it before the doc becomes finalized or before a hardfork EIP specifying it

File: EIPS/eip-5283.md
Comment (xinbenlv) [inline EIPS/eip-5283.md:32]: Same here

File: EIPS/eip-5283.md
Comment (xinbenlv) [inline EIPS/eip-5283.md:70]:
```suggestion
1. Transient storage opcodes (`TLOAD`/`TSTORE`) provide contract state that is kept between calls in the same transaction but it is not committed to the world state afterward. These opcodes also enable fine-grained parallelization. 
```

File: EIPS/eip-5283.md
Comment (xinbenlv) [inline EIPS/eip-5283.md:74]:
```suggestion
All these alternative proposals have the downside that they create new opcodes, and this is discouraged if the same functionality can be implemented with the same gas cost using precompiles. A new opcode requires modifying compilers, debuggers and static analysis tools.
```

File: EIPS/eip-5283.md
Comment (xinbenlv) [inline EIPS/eip-5283.md:34]: IMHO this is more suitable to be in the reference implementation (of usage)

File: EIPS/eip-5283.md
Comment (xinbenlv) [inline EIPS/eip-5283.md:82]: [Optional] Could you further provide your suggestion of how much the gas cost would you suggest?

**Review by xinbenlv — DISMISSED** [review]
LGTM with nitpicks

File: EIPS/eip-5283.md
Comment (Pandapip1) [inline EIPS/eip-5283.md:90]:
```suggestion
```solidity
```
Nit.

**Review by Pandapip1 — CHANGES_REQUESTED** [review]
@xinbenlv had some good comments too. I recommend putting placeholders like `CONTRACT_ADDRESS` instead of `0x0A`, and things like `FORK_BLKNUM` for when the precompiled address is to be deployed.

**Review by Pandapip1 — APPROVED** [review]

**Comment by Pandapip1** [comment]
This is fine as-is for a draft but needs some serious changes before going to review.

**Pull request merged**
